lib/internal/source_map/source_map_cache.js

```diff
@@ -0,0 +1,196 @@
+        sourceMappingURL: match.groups.sourceMappingURL,
+        payload: dataFromUrl(basePath, match.groups.sourceMappingURL)
+      });
+    } else {
+      // If there is no cjsModuleInstance assume we are in a
+      // "modules/esm" context.
+      esmSourceMapCache.set(filename, {
+        sourceMappingURL: match.groups.sourceMappingURL,
+        payload: dataFromUrl(basePath, match.groups.sourceMappingURL)
+      });
+    }
+  }
+}
+
+function dataFromUrl(basePath, sourceMappingURL) {
+  try {
+    const url = new URL(sourceMappingURL);
+    switch (url.protocol) {
+      case 'data:':
+        return sourceMapFromDataUrl(basePath, url.pathname);
+      default:
+        debug(`unknown protocol ${url.protocol}`);
+        return null;
+    }
+  } catch (err) {
+    debug(err.stack);
+    // If no scheme is present, we assume we are dealing with a file path.
+    const sourceMapFile = resolve(basePath, sourceMappingURL);
+    return sourceMapFromFile(sourceMapFile);
+  }
+}
+
+function sourceMapFromFile(sourceMapFile) {
+  try {
+    const content = fs.readFileSync(sourceMapFile, 'utf8');
+    const data = JSON.parse(content);
+    return sourcesToAbsolute(dirname(sourceMapFile), data);
+  } catch (err) {
+    debug(err.stack);
+    return null;
+  }
+}
+
+// data:[<mediatype>][;base64],<data> see:
+// https://tools.ietf.org/html/rfc2397#section-2
+function sourceMapFromDataUrl(basePath, url) {
+  const [format, data] = url.split(',');
+  const splitFormat = format.split(';');
+  const contentType = splitFormat[0];
+  const base64 = splitFormat[splitFormat.length - 1] === 'base64';
+  if (contentType === 'application/json') {
+    const decodedData = base64 ?
+      Buffer.from(data, 'base64').toString('utf8') : data;
+    try {
+      const parsedData = JSON.parse(decodedData);
+      return sourcesToAbsolute(basePath, parsedData);
+    } catch (err) {
+      debug(err.stack);
+      return null;
+    }
+  } else {
+    debug(`unknown content-type ${contentType}`);
+    return null;
+  }
+}
+
+// If the sources are not absolute URLs after prepending of the "sourceRoot",
+// the sources are resolved relative to the SourceMap (like resolving script
+// src in a html document).
+function sourcesToAbsolute(base, data) {
+  data.sources = data.sources.map((source) => {
+    source = (data.sourceRoot || '') + source;
+    if (!/^[\\/]/.test(source[0])) {
+      source = resolve(base, source);
+    }
+    if (!source.startsWith('file://')) source = `file://${source}`;
+    return source;
+  });
+  // The sources array is now resolved to absolute URLs, sourceRoot should
+  // be updated to noop.
+  data.sourceRoot = '';
+  return data;
+}
+
+// Move source map from garbage collected cache to cache keyed on filename.
+function persistSourceMap(filename) {
+  const sourceMap = getSourceMap(filename);
+  if (sourceMap) {
+    setSourceMap(
+      filename,
+      sourceMap.sourceMappingUrl,
+      sourceMap.payload
+    );
+  }
+}
```

Comment:
This is problematic. Are we freeing it from the cache anytime?

Reply:
what if we put a flag on the entry if it's persisted for the purposes of an exceptional flow, e.g.,

`persitForException`, with an additional parameter, `exceptional: true`, and then when we decorate the stack trace we remove it from cache?

my only concern is the edge-case that an exception happens, and will cause the process to exit, we want that information on disk for source-maps ... will we know by the time that `prepareStackTrace` is called whether the process is slated to be destroyed (@devsnek you might know?).

Reply:
That would be the way to go.

Reply:
you could flag to `prepareStackTrace` whether it was called while formatting a fatal exception, but you still run into a problem:
```js
const e = new Error();
e.stack; // generate stack before fatal exception logic
throw e;
```
not that i think anyone does this *directly*, but it may happen indirectly from custom formatting logic, random console.logs, etc.

Reply:
Given that errors are often wrapped in Promise.reject (ie, not always thrown) it seems pretty likely.

Reply:
I figured out an approach that works for most scenarios; simply move the source-map to being keyed on `error` rather than `module` if compile throws.